File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ProfileHeaderView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "ProfileHeaderView.h"

#import "WAZUIMagicIOS.h"
#import "UIImage+ZetaIconsNeue.h"
#import <PureLayout/PureLayout.h>

@import WireExtensionComponents;
#import <PureLayout/PureLayout.h>
#import "WireStyleKit.h"
#import "Wire-Swift.h"


@interface ProfileHeaderView ()

@property (nonatomic, assign) ProfileHeaderStyle headerStyle;
@property (nonatomic, strong, readwrite) UILabel *titleLabel;
@property (nonatomic, strong, readwrite) UILabel *akaLabel;
@property (nonatomic, strong) UIImageView *akaImageView;
@property (nonatomic, strong) UIImageView *verifiedImageView;
@property (nonatomic, strong, readwrite) UITextView *subtitleLabel;
@property (nonatomic, strong, readwrite) IconButton *dismissButton;
@property (nonatomic) NSLayoutConstraint *subtitleLabelTopOffset;
@end



@implementation ProfileHeaderView

- (instancetype)initWithHeaderStyle:(ProfileHeaderStyle)headerStyle {
    // Use non-zero rect to avoid broken autolayout
	if (self = [super initWithFrame:CGRectMake(0, 0, 320, 80)]) {
		_headerStyle = headerStyle;
		[self createViews];
		[self setupConstraints];
	}
	return self;
}

- (void)createViews
{
    self.titleLabel = [[UILabel alloc] initForAutoLayout];
    [self addSubview:self.titleLabel];
    
    self.akaLabel = [[UILabel alloc] initForAutoLayout];
    self.akaLabel.backgroundColor = [UIColor clearColor];
    [self addSubview:self.akaLabel];
    
    self.akaImageView = [[UIImageView alloc] initWithImage:[UIImage imageForIcon:ZetaIconTypeAddressBook
                                                                        iconSize:ZetaIconSizeSearchBar
                                                                           color:[UIColor colorWithMagicIdentifier:@"style.color.foreground.faded"]]];
    [self addSubview:self.akaImageView];
    
    self.verifiedImageView = [[UIImageView alloc] initWithImage:[WireStyleKit imageOfShieldverified]];
    self.verifiedImageView.accessibilityIdentifier = @"VerifiedShield";
    [self addSubview:self.verifiedImageView];
    self.verifiedImageView.hidden = YES;

    self.subtitleLabel = [[LinkInteractionTextView alloc] initForAutoLayout];
	self.subtitleLabel.editable = NO;
    self.subtitleLabel.scrollEnabled = NO;
	self.subtitleLabel.textContainerInset = UIEdgeInsetsZero;
	self.subtitleLabel.textContainer.lineFragmentPadding = 0;
	self.subtitleLabel.textContainer.maximumNumberOfLines = 1;
	self.subtitleLabel.textContainer.lineBreakMode = NSLineBreakByTruncatingTail;
	self.subtitleLabel.dataDetectorTypes = UIDataDetectorTypeLink;
    self.subtitleLabel.backgroundColor = [UIColor clearColor];
    [self addSubview:self.subtitleLabel];
    
    self.dismissButton = [IconButton iconButtonCircular];
    self.dismissButton.translatesAutoresizingMaskIntoConstraints = NO;
    self.dismissButton.accessibilityIdentifier = @"OtherUserProfileCloseButton";
    

	switch (self.headerStyle) {
		case ProfileHeaderStyleBackButton: {
            [self.dismissButton setIcon:ZetaIconTypeChevronLeft withSize:ZetaIconSizeTiny forState:UIControlStateNormal];
			break;
		}

		case ProfileHeaderStyleCancelButton: {
            [self.dismissButton setIcon:ZetaIconTypeX withSize:ZetaIconSizeTiny forState:UIControlStateNormal];
			break;
		}

		case ProfileHeaderStyleNoButton:
			self.dismissButton.hidden = YES;
			break;
	}

    [self addSubview:self.dismissButton];
}

- (void)setupConstraints
{

	CGFloat contentTopMargin = [WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_top_margin"];

    [self.titleLabel autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:contentTopMargin];
	[self.titleLabel autoSetDimension:ALDimensionHeight toSize:32];
    [self.titleLabel autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:[WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_left_margin"] + 32];
    [self.titleLabel autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:[WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_left_margin"] + 32];

    [self.akaLabel addConstraintForAligningTopToBottomOfView:self.titleLabel distance:4];
    [self.akaLabel autoAlignAxisToSuperviewAxis:ALAxisVertical];
    
    [self.akaImageView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:self.akaLabel];
    [self.akaImageView autoPinEdge:ALEdgeLeft toEdge:ALEdgeRight ofView:self.akaLabel withOffset:8];
    [self.akaLabel autoPinEdge:ALEdgeRight toEdge:ALEdgeLeft ofView:self.akaImageView withOffset:-8];
    
    self.subtitleLabelTopOffset = [self.subtitleLabel autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.akaLabel withOffset:10];
    [self.subtitleLabel autoAlignAxisToSuperviewAxis:ALAxisVertical];
    [self.subtitleLabel autoPinEdgeToSuperviewEdge:ALEdgeBottom];
	[self.subtitleLabel autoSetDimension:ALDimensionHeight toSize:32];

	CGFloat dismissButtonTopMargin = 24;

    [self.dismissButton autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:dismissButtonTopMargin];
    [self.dismissButton autoMatchDimension:ALDimensionWidth toDimension:ALDimensionHeight ofView:self.dismissButton];
    [self.dismissButton autoSetDimension:ALDimensionWidth toSize:32];
    
    [self.verifiedImageView autoAlignAxis:ALAxisHorizontal toSameAxisOfView:self.titleLabel];
    [self.verifiedImageView autoMatchDimension:ALDimensionWidth toDimension:ALDimensionHeight ofView:self.verifiedImageView];
    [self.verifiedImageView autoSetDimension:ALDimensionWidth toSize:16];
    [self.verifiedImageView autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:[WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_left_margin"]];
    
    
	switch (self.headerStyle) {
		case ProfileHeaderStyleBackButton:
			[self.dismissButton autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:[WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_left_margin"]];

			break;
		case ProfileHeaderStyleCancelButton:
            [self.dismissButton autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:[WAZUIMagic cgFloatForIdentifier:@"profile_temp.content_right_margin"]];
			break;

		case ProfileHeaderStyleNoButton:

			break;
	}
}

 - (void)setShowVerifiedShield:(BOOL)showVerifiedShield
{
    _showVerifiedShield = showVerifiedShield;
    BOOL shouldHide = YES;
    if (self.headerStyle != ProfileHeaderStyleBackButton) {
        shouldHide = !showVerifiedShield;
    }
    
    [UIView transitionWithView:self.verifiedImageView
                      duration:0.2
                       options:UIViewAnimationOptionTransitionCrossDissolve
                    animations:^{
        self.verifiedImageView.hidden = shouldHide;
    } completion:nil];
}

- (CGSize)intrinsicContentSize
{

    if (self.subtitleLabel.text.length != 0) {
        return CGSizeMake(UIViewNoIntrinsicMetric, self.subtitleLabel.bounds.size.height);
    }
    else if (!self.dismissButton.hidden) {
        return CGSizeMake(UIViewNoIntrinsicMetric, self.dismissButton.bounds.size.height);
    }
    else if (self.titleLabel.text.length != 0) {
        return CGSizeMake(UIViewNoIntrinsicMetric, self.titleLabel.bounds.size.height);
    }
    else {
        return CGSizeMake(UIViewNoIntrinsicMetric, 30.0f);
    }
}

- (void)updateConstraints
{
    [self invalidateIntrinsicContentSize];
    self.akaImageView.hidden = (self.akaLabel.attributedText.length == 0);
    self.subtitleLabelTopOffset.constant = (self.akaLabel.attributedText.length == 0) ? 0 : 10;
    [super updateConstraints];
}

@end

File: Wire-iOS/Sources/UserInterface/UserProfile/Views/ProfileHeaderView.swift
//
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit
import Cartography

enum ProfileHeaderStyle: Int {
    case cancelButton, backButton, noButton
}

final class ProfileHeaderView: UIView {

    var showVerifiedShield = false {
        didSet {
            updateVerifiedShield()
        }
    }

    private(set) var dismissButton = IconButton.iconButtonCircular()
    internal(set) var headerStyle: ProfileHeaderStyle
    /// flag for disable headerStyle update in traitCollectionDidChange. It should be used for test only.
    private let navigationControllerViewControllerCount: Int?
    private let profileViewControllerContext: ProfileViewControllerContext?

    private let detailView = UserNameDetailView()
    private let verifiedImageView = UIImageView(image: WireStyleKit.imageOfShieldverified())

    private var backButtonLeading: NSLayoutConstraint?
    private var cancelButtonTrailing: NSLayoutConstraint?

    @objc(initWithViewModel:)
    init(with viewModel: ProfileHeaderViewModel) {
        headerStyle = .noButton
        navigationControllerViewControllerCount = viewModel.navigationControllerViewControllerCount
        profileViewControllerContext = viewModel.context
        super.init(frame: .zero)

        setupViews()
        configure(with: viewModel)
        createConstraints()
        updateHeaderStyle()
        updateDismissButton()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupViews() {
        translatesAutoresizingMaskIntoConstraints = false

        [detailView, dismissButton, verifiedImageView].forEach(addSubview)
        verifiedImageView.isHidden = true

        verifiedImageView.accessibilityIdentifier = "VerifiedShield"
        dismissButton.accessibilityIdentifier = "OtherUserProfileCloseButton"
    }

    private func createConstraints() {
        let topMargin = WAZUIMagic.cgFloat(forIdentifier: "profile_temp.content_top_margin")
        let horizontalMargin = WAZUIMagic.cgFloat(forIdentifier: "profile_temp.content_left_margin")

        let detailViewMargin = horizontalMargin + 32

        constrain(self, detailView) { view, detailView in
            detailView.top == view.top + topMargin
            detailView.leading == view.leading + detailViewMargin
            detailView.trailing == view.trailing - detailViewMargin
            detailView.bottom == view.bottom - 12
        }

        constrain(self, dismissButton, verifiedImageView, detailView.titleLabel) { view, dismiss, verified, title in
            dismiss.top == view.top + 26
            dismiss.width == dismiss.height
            dismiss.width == 32

            verified.centerY == title.centerY
            verified.width == verified.height
            verified.width == 16
            verified.leading == view.leading + horizontalMargin

            backButtonLeading = dismiss.leading == view.leading + horizontalMargin
            cancelButtonTrailing = dismiss.trailing == view.trailing - horizontalMargin
        }
    }

    @objc(configureWithViewModel:)
    public func configure(with model: ProfileHeaderViewModel) {
        detailView.configure(with: model.userDetailViewModel)
    }

    private func updateVerifiedShield() {
        var shouldHide = true
        if headerStyle != .backButton {
            shouldHide = !showVerifiedShield
        }

        UIView.transition(
            with: verifiedImageView,
            duration: 0.2,
            options: .transitionCrossDissolve,
            animations: { self.verifiedImageView.isHidden = shouldHide },
            completion: nil
        )
    }

    // MARK: - DismissButton style and constraints

    func updateDismissButton() {
        switch headerStyle {
        case .backButton:
            cancelButtonTrailing?.isActive = false
            backButtonLeading?.isActive = true

            dismissButton.isHidden = false

            dismissButton.setIcon(.chevronLeft, with: .tiny, for: .normal)
        case .cancelButton:
            backButtonLeading?.isActive = false
            cancelButtonTrailing?.isActive = true

            dismissButton.isHidden = false

            dismissButton.setIcon(.X, with: .tiny, for: .normal)
        case .noButton:
            ///the button is hidden, the position is not important
            backButtonLeading?.isActive = false
            cancelButtonTrailing?.isActive = true

            dismissButton.isHidden = true
        }
    }

    private func updateHeaderStyle() {
        var headerStyle: ProfileHeaderStyle = .cancelButton

        if self.traitCollection.userInterfaceIdiom == .pad && UIApplication.shared.keyWindow?.traitCollection.horizontalSizeClass == .regular {

            if navigationControllerViewControllerCount > 1 {
                headerStyle = .backButton
            } else if profileViewControllerContext != .deviceList {
                headerStyle = .noButton
            }
        }

        self.headerStyle = headerStyle
    }

    // MARK: - UITraitCollection

    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        guard self.traitCollection.userInterfaceIdiom == .pad,
            UIApplication.shared.keyWindow?.traitCollection.horizontalSizeClass != .unspecified else {
            return
        }

        updateHeaderStyle()
        updateDismissButton()
    }
}

